internal/checker/nodebuilderimpl.go

```diff
@@ -0,0 +1,2972 @@
+}
+
+func (b *nodeBuilderImpl) createExpressionFromSymbolChain(chain []*ast.Symbol, index int) *ast.Expression {
+	// typeParameterNodes := b.lookupTypeParameterNodes(chain, index)
+	symbol := chain[index]
+
+	if index == 0 {
+		b.ctx.flags |= nodebuilder.FlagsInInitialEntityName
+	}
+	symbolName := b.getNameOfSymbolAsWritten(symbol)
+	if index == 0 {
+		b.ctx.flags ^= nodebuilder.FlagsInInitialEntityName
+	}
+
+	if startsWithSingleOrDoubleQuote(symbolName) && core.Some(symbol.Declarations, hasNonGlobalAugmentationExternalModuleSymbol) {
+		return b.f.NewStringLiteral(b.getSpecifierForModuleSymbol(symbol, core.ResolutionModeNone))
+	}
+
+	if index == 0 || canUsePropertyAccess(symbolName, b.ch.languageVersion) {
+		identifier := b.f.NewIdentifier(symbolName)
+		b.e.AddEmitFlags(identifier, printer.EFNoAsciiEscaping)
+		// !!! TODO: smuggle type arguments out
+		// if (typeParameterNodes) setIdentifierTypeArguments(identifier, factory.createNodeArray<TypeNode | TypeParameterDeclaration>(typeParameterNodes));
+		// identifier.symbol = symbol;
+		if index > 0 {
+			return b.f.NewPropertyAccessExpression(b.createExpressionFromSymbolChain(chain, index-1), nil, identifier, ast.NodeFlagsNone)
+		}
+		return identifier
+	}
+
+	if startsWithSquareBracket(symbolName) {
+		symbolName = symbolName[1 : len(symbolName)-1]
+	}
+
+	var expression *ast.Expression
+	if startsWithSingleOrDoubleQuote(symbolName) && symbol.Flags&ast.SymbolFlagsEnumMember == 0 {
+		expression = b.f.NewStringLiteral(unquoteString(symbolName))
+	} else if jsnum.FromString(symbolName).String() == symbolName {
+		// TODO: the follwing in strada would assert if the number is negative, but no such assertion exists here
+		// Moreover, what's even guaranteeing the name *isn't* -1 here anyway? Needs double-checking.
+		expression = b.f.NewNumericLiteral(symbolName)
+	}
+	if expression == nil {
+		expression = b.f.NewIdentifier(symbolName)
+		b.e.AddEmitFlags(expression, printer.EFNoAsciiEscaping)
+		// !!! TODO: smuggle type arguments out
+		// if (typeParameterNodes) setIdentifierTypeArguments(identifier, factory.createNodeArray<TypeNode | TypeParameterDeclaration>(typeParameterNodes));
+		// identifier.symbol = symbol;
+		// expression = identifier;
+	}
+	return b.f.NewElementAccessExpression(b.createExpressionFromSymbolChain(chain, index-1), nil, expression, ast.NodeFlagsNone)
+}
+
+func canUsePropertyAccess(name string, languageVersion core.ScriptTarget) bool {
+	if len(name) == 0 {
+		return false
+	}
+	// TODO: in strada, this only used `isIdentifierStart` on the first character, while this checks the whole string for validity
+	// - possible strada bug?
+	if strings.HasPrefix(name, "#") {
+		return len(name) > 1 && scanner.IsIdentifierText(name[1:], languageVersion, core.LanguageVariantStandard)
+	}
+	return scanner.IsIdentifierText(name, languageVersion, core.LanguageVariantStandard)
+}
+
+func unquoteString(str string) string {
```

Comment:
This would be good to shift into `stringutil` or something; I suspect we have it (or will need it for something else in the future).